### ISSUE TITLE
Feature: JWT token with AuthorizationHandler for dynamic authentication

### ISSUE THREAD

**Issue by oliverhausler** [nats-io/nats-streaming-server#772]
Is an `AuthorizationHandler` on the roadmap?

I just read the docs, and there seem to be these password modes:

- user/password
- token
...with permissions for specific subjects per user/token

As I understand, this is part of the static server configuration. Is the configuration hot-reloaded, e.g. could I add and remove permissions at runtime? Probably not.

Then my suggestion would be an `AuthorizationHandler`, where my own code can receive the token which the client passed in during connect, validate it and match subjects against it. More specifically, I would like to have my authentication server generate JWT tokens which contain allowed and/or disallowed subjects, then give these to clients, which use them against NATS.

Is there are reason why there are no custom handlers? Performance?

**Comment by kozlovic** [comment]
We need to work on auth/permissions in NATS Streaming. By embedding NATS Server, we can leverage the NATS auth/permissions, but it is tricky to limit an user to a specific set of channels and more specifically to prevent subscriptions.
At this point, NATS Streaming vendors NATS Server 1.4.1, for which there is no JWT support (NATS 2.0.0 - not released yet - has a lot of enhancements in term of auth/permissions, etc..).

**Comment by oliverhausler** [comment]
Ivan, thanks. Will my use case be covered in NATS 2.0.0, and is there a proposed ETA? We are planning a new development based on `nats-streaming`, and if we'll see NATS 2.0.0 in about six months, I'd be more than happy. 

_I'll post another issue with more feature requests, so you have these, too._

**Comment by kozlovic** [comment]
We are not sure yet if we will be able to leverage NATS 2.0.0 to solve all auth/perm issues for NATS Streaming. But definitively send requirements so that we can have that in mind for Streaming..

**Comment by oliverhausler** [comment]
So my suggestion would be this:

- Our auth server creates a JWT with a user ID, a string array or allowed and denied subjects, and an expiration date:
```
{
  "sub": "userId",
  "subjects" : {
    "allow": {"", "", ""},
    "deny": {}
  },
  "exp": 1516239022
}
```
- The JWT is signed by our auth server, using a `secret`, which only our server and the NATS server knows.

Then on NATS:

- The `secret` is specified in config or passed as env variable.
- When the client connects, the signature in the JWT is verified against the `secret` to ensure authenticity.
- If authentic, allowed and denied subjects, as well as the expiration date is stored for the connection.
- When the connection subscribes a subject, NATS checks if the subject is allowed and the JWT has not expired.

NB: It doesn't make sense to specify `allow` and `deny` at the same time, but we could allow positive and negative logic if we wanted.

**Comment by mrparkers** [comment]
I am also interested in JWT based authentication for nats-streaming.  I agree with the requirements that @oliverhausler has listed above, however I would like to add an additional requirement for supporting asymmetrically signed JWTs.

The key difference is that asymmetrically signed JWTs do not require the NATs streaming server to be aware of a pre-shared secret.  Instead, when the NATs streaming server receives a JWT, it can download a public key from a well-known endpoint ([example](https://example.auth0.com/.well-known/jwks.json)) and verify the JWT's signature that way.  This allows for an extra layer of security by not providing the NATs server with a key that it could, in theory, use to sign any JWT it wants.  It also allows us to rotate JWKs without requiring a configuration change to the NATs server.

I also want to echo the original request to allow custom code to be used for authentication/authorization purposes, similar to the `CustomClientAuthentication` for `gnatsd`.  If my understanding is correct (please correct me if I am wrong here), `CustomClientAuthentication` can't be used for this purpose because that only controls custom authentication logic for clients connecting to the embedded NATs server, which is redundant because the NATs streaming server is the only client.  Thus, having a configuration option like this for the NATs streaming server would at least allow me to implement my own JWT verification logic if this feature ends up getting deprioritized.

**Comment by wutkemtt** [comment]
We also need a way to configure a JWT client token for connecting nats streaming to an external nats server. Due to the usage of leafnodes we had to use the JWTs for authentication in combination with the nats account server. The problem is, when the nats server is running with JWT authentication, whether a connection with username and password nor token could be made. These are the only authentication types nats streaming is supporting for now. That means we have to decide whether to use the required nats streaming or the leafnodes concept.

**Comment by kozlovic** [comment]
@wutkemtt The latest NATS Streaming server was updated to use NATS 2.0 embedded server due to some pressing issues with go.mod story, but it is correct that the product has not been updated yet to allow full 2.0 features.

**Comment by walfly** [comment]
@kozlovic I'm also looking for very similar functionality to whats described above.  Is this still on the roadmap? any update on how far off it is?

**Comment by kozlovic** [comment]
Since NATS Streaming v0.16.2, there is an option to provide the credentials file for the Streaming server to use to connect to a NATS Server 2.x. You can use a JWT and connect to a NATS Server or through a leafnode. But as @mrparkers mentioned, this is just for the NATS Streaming server (that is a really a client) connecting to a NATS Server. Streaming clients can create their own NATS  connection using JWT to connect to the NATS Server, but again, the issue is with permissions and NATS Streaming, or the lack thereof. Implementing all the features of NATS 2.0 at the level of Streaming is not something we are planning on doing. Instead, JetStream (available soon and already in preview/beta) addresses those issue because it is all account based and part of the core NATS Server. So you won't need specific client libraries (just core) to use streaming/persistence with JetStream.

I am closing this issue so that people are not under the impression that this is under development. Thank you for your understanding.

**Comment by kozlovic** [comment]
By the way, the configuration is "credentials" in the streaming section of the config: https://docs.nats.io/nats-streaming-server/configuring/cfgfile.

**Comment by alexykot** [comment]
@kozlovic So the issue of auth-n/z will not be addressed in the NATS Streaming Server, which is understandable since JetStream is meant to address this and many other issues. However, JetStream is still in the tech preview state and I don't think I can use it now in my current project.

My strong preference is to stay within Go ecosystem if possible, so I'm considering the option to implement custom auth-n/z to augment NATS Streaming Server. For core NATS I see `nopts.CustomClientAuthentication` server option and corresponding `Authentication` interface that I could implement and add features I need without much intrusion in the server code itself. However, I can't find anything similar in the NATS Streaming options or server code.

Is there a relatively simple way to add custom authentication handler for the NATS Streaming Server?

**Comment by oliverhausler** [comment]
@alexykot I don't know if it covers what you need, but you may want to take a look at https://github.com/liftbridge-io/liftbridge.

**Comment by alexykot** [comment]
Yeah, that's another option we're considering :)

Although, tbh that project looks a little niche and far not as developed and adopted as NATS. But maybe it's just my impressing from reading the docs.

**Comment by kozlovic** [comment]
@alexykot This original issue is old and since then the streaming server does embed NATS 2.1.9, which has NATS 2.0 auth features. Note that the configuration reload works in streaming for the NATS related configuration. It is always possible to run a separate NATS Server anyway and have streaming connect to it.
NATS Streaming can connect to the NATS server using credentials, user/pass, token, etc..
What exactly are you looking for?

**Comment by alexykot** [comment]
I'm looking for an ability to control streaming publisher and consumer access individually per-channel and per-subscriber. I'm not looking for controlling access of the Streaming Server itself to the NATS Server, I need to control access of Steaming Server's clients.

In some other issue I think I've seen a note that consumer access control specifically is not possible via NATS auth, but I've not saved a link to it.

**Comment by kozlovic** [comment]
@alexykot 

> I'm looking for an ability to control streaming publisher and consumer access individually per-channel and per-subscriber. I'm not looking for controlling access of the Streaming Server itself to the NATS Server, I need to control access of Steaming Server's clients.

For producers, this is doable. See https://github.com/nats-io/nats-server/issues/1713 for reference.

> In some other issue I think I've seen a note that consumer access control specifically is not possible via NATS auth, but I've not saved a link to it.

That is correct. Subscriptions are created by clients sending a protocol to a subject such as `_STAN.sub.<nuid/cluster id`. The actual subject the client wants to subscribe to is inside the protocol. There have been talks about adding a `_STAN.sub.<nuid/cluster id>.>` and make the end of the subject the actual subject the client want to subscribe. This would allow permissions to work the same way that you can for publish protocols. However, this would require changes in client and server and potentially incompatibility..

Are you planning on embedding the NATS Streaming server? Since you mentioned the idea of an auth/perm handler I guess that is what you are planning on doing. We could certainly add a new option in Streaming that is invoked when processing a subscription request.

**Comment by alexykot** [comment]
Yes, I'm looking to embed the streaming server and add needed auth-n/z
mechanics to it.

I would prefer this to be a thin wrapper with custom authenticator
interface supplied on server bootstrap. But by the looks of it, I will have
to fork the stan server itself and customize it, since current codebase
does not provide a way to supply custom authenticator.

I'm not sure that is feasible with my current resources tbh, so if it's
possible to expose the needed interface as a server option in the upstream
- that would be great.

**Comment by kozlovic** [comment]
@alexykot What interfaces you would be interested in?
- user auth (server would provide you with client ID)?
- publish permission (server would provide the client ID the publish comes from and channel message is published to)?
- subscription request (server would provide the client ID and the channel the subscription is for)?

The first and last seem ok, the publish permission may be costly since it would be per message.

So you would specify auth handlers in the options and if set, the server would invoke them based on events described above. Is this what you had in mind?

**Comment by alexykot** [comment]
Hmm, I'm reaching the limits of my current knowledge of NATS internal architecture, so sorry if my response will not make a lot of sense, but I'll try.

In principle, in my naive view, I imagine it to work something along these lines:
1. The stan client library accepts an opaque auth token (assume JWT) and forwards it to the stan server.
2. The stan server receives the request and invokes the auth handler, providing the token, client ID and channel ID to it.
3. Auth handler validates the token and returns true/false for given token/client ID/channel ID.

How the tokens are produced, delivered to clients and validated by the auth handler is out of scope - we can handle that on our side, bootstrapping some rudimentary PKI and supplying auth handlers with correct public keys for token validation. The token will carry the permission scopes and the client ID.

So, yeah, what you describe makes sense, we totally can provide custom auth handlers for these actions and validate users. Authentication will be done by the token validation and authorisation will be done using permission scopes carried in the token. As long as we can provide, forward and check the tokens - it should work.

The per-message validation on publishing may be an issue, but I see how we could potentially optimise that in our case. For example - I imagine hashing and caching known good tokens will be much faster than explicitly validating it on every message, and tokens won't change that often. So I wouldn't worry about that too much right now, this performance constraint is outside the server itself and up to the auth handler to manage, and in my use case, the reliable auth takes priority over performance.

**Comment by alexykot** [comment]
Btw, a bit off-topic, but I've also posted two questions on NATS Streaming Server at StackOverflow over here:
https://stackoverflow.com/questions/64800525/how-many-channels-i-can-have-with-nats-streaming-server
https://stackoverflow.com/questions/64800182/what-message-persistence-guarantee-nats-streaming-provides-in-cluster-and-ft-mod
Not sure if that's the best place, or reposting it here on GitHub would be better.

**Comment by kozlovic** [comment]
@alexykot I did post an answer to your SO questions. Personally, I prefer people to post here on GH so I have a single place where all issues/questions are asked.

Regarding JWT, I thought about it more and re-reading everything that was posted here, and again, what you are trying to do is exactly what NATS 2.0 already does. We have support for JWT (with permissions included) with NKeys support. The streaming library already allows you to specify the NATS JWT through credentials file. You then pass the NATS connection to the STAN connection. The only thing that would not work is because of the way the server handles subscription requests (as I mentioned earlier). Should the server listens to `_STAN.sub.[nuid or cluster name].>` instead of `_STAN.sub.[nuid or cluster name]` and the client send the request with the suffix being the channel it wants to subscribe to, then admin could limit that with permissions for users on say `_STAN.sub.*.foo`, etc..

So if anything we should probably go this direction instead of auth handler with API in the clients to pass a JWT, which may be confusing for users already using NATS JWTs. Furthermore, the approach you mentioned would work only for users embedding NATS Streaming server (so they can implement the handlers) while the approach with changing subject for subscriptions request would work for executables.
To be fair, I have not tested this and will do this when I find the time. But if this work, and with the subscription subject change, it would then allow users to use NATS JWT permissions to limit connecting, publishing (as it is already the case now) and subscribing (which is missing).

**Comment by alexykot** [comment]
> I did post an answer to your SO questions. Personally, I prefer people to post here on GH so I have a single place where all issues/questions are asked.
Thanks, noted. 

Ok, I will look again at NATS2.0 auth model and will try to test it together with Streaming Server the way you describe.

**Comment by kozlovic** [comment]
To be clear, since currently the clients send subscriptions request without the subject as part of the subscription request subject, you would not be able to limit with permissions the client ability to subscribe. This would still require a change in the client library and server.

**Comment by alexykot** [comment]
Yes, I get that in principle. 

I just want to better understand NATS 2.0 auth model and the interaction between NATS and Streaming Server before coming back with more questions.

**Comment by alexykot** [comment]
So I've got to study NATS auth-n/z, NATS subject mechanics and STAN's use of them, and made a prototype for what we want to achieve. 

I now understand what you mean about `_STAN.pub.*.foo` and `_STAN.sub.*` subjects. Yes, it's possible to control STAN stream publish access by controlling publish access to corresponding `_STAN.pub.*.foo` NATS subject, where `foo` is the STAN stream name. But it is not possible to control the subscribe access this way because STAN subscriptions are creating NATS subjects with names `_STAN.sub.*`, without `foo` suffix, so we cannot enforce access basing on STAN stream names here.

Question - how hard would it be to change this behaviour of STAN server and client to include the `foo` STAN stream name into NATS subject?

**Comment by kozlovic** [comment]
@alexykot Good to know that you see how permissions on _STAN subjects could be used to solve what you are trying to do. The missing part would be for subscriptions. It is not that difficult to make the changes but requires changes in clients and server. The server on connect could return an info indicating to the client library that it uses the new subscription request subject and new clients supporting that would send to the new subject.

The main issue I see is that if the servers (say in a cluster deployment) are not at the same level of version/configuration. A newer client that connects first to a server with this ability will have no way of knowing if it later sends a subscription to an older server (which subscription request would timeout). This is because the "connect" handshake is done only once.

I am busy at the moment with NATS Server and MQTT, but may have more free time in the coming weeks to look at this and other things I want to add/fix in NATS Streaming.

**Comment by alexykot** [comment]
Ok, good to know that this is not a complicated task. We're only evaluating streaming system options right now and it's a few months before first iteration of NATS-based architecture will hit production, so if we will have this fix then - would be great.

Otherwise, as a backstop solution, since it sounds like it's not a big task - we may be able to fix it ourselves, PR it and use a fork until it's fixed in the upstream. Ofc we don't have some of the concerns you have, e.g. around making old and new way play nicely in a single cluster, we can workaround that in our deployment.
